##UJIAN TENGAH SEMESTER DATA MINING
#Prediksi Kualitas Anggur (Wine Quality)
Nama: (Widia Pragustin) NIM: (2304020045) Rombel: R1 Pendidikan Matematika

## 1. Import Library
Pada tahap ini dilakukan import library yang dibutuhkan untuk analisis data dan pembuatan model.

In [1]:
import pandas as pd
import numpy as np

# visualisasi
import matplotlib.pyplot as plt
import seaborn as sns

# preprocessing & model
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# model
from sklearn.ensemble import RandomForestClassifier

# evaluasi
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

Penjelasan:
- pandas → untuk membaca dan mengolah data
- numpy → untuk perhitungan numerik
- matplotlib & seaborn → untuk visualisasi data
- sklearn → untuk machine learning

## 2. Load Dataset
Dataset yang digunakan adalah dataset Wine Quality yang berisi fitur kimia dan kualitas anggur.

In [4]:
train = pd.read_csv('data_training.csv')
test = pd.read_csv('data_testing.csv')

In [5]:
train.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
0,7.3,0.740,0.08,1.7,0.094,10.0,45.0,0.99576,3.24,0.50,9.8,5,1366
1,8.1,0.575,0.22,2.1,0.077,12.0,65.0,0.99670,3.29,0.51,9.2,5,103
2,10.1,0.430,0.40,2.6,0.092,13.0,52.0,0.99834,3.22,0.64,10.0,7,942
3,12.9,0.500,0.55,2.8,0.072,7.0,24.0,1.00012,3.09,0.68,10.9,6,811
4,8.4,0.360,0.32,2.2,0.081,32.0,79.0,0.99640,3.30,0.72,11.0,6,918


- Dataset training digunakan untuk melatih model
- Dataset testing digunakan untuk prediksi
- Fungsi head() digunakan untuk melihat 5 data pertama

### Interpretasi Data Awal

Berdasarkan output `train.head()`, terlihat bahwa dataset training berisi beberapa variabel numerik yang merepresentasikan karakteristik kimia dari anggur.

Beberapa fitur yang terdapat dalam dataset antara lain:
- fixed acidity → tingkat keasaman tetap
- volatile acidity → keasaman yang mudah menguap
- citric acid → kandungan asam sitrat
- residual sugar → kadar gula tersisa
- chlorides → kadar garam
- free sulfur dioxide → sulfur dioksida bebas
- total sulfur dioxide → total sulfur dioksida
- density → massa jenis
- pH → tingkat keasaman
- sulphates → kandungan sulfat
- alcohol → kadar alkohol

Selain itu terdapat:
- quality → sebagai variabel target (yang akan diprediksi)
- Id → sebagai identitas data

Dari tampilan 5 data pertama, terlihat bahwa semua data berbentuk numerik sehingga siap untuk dilakukan analisis lebih lanjut.

## 3. Memahami Struktur Data

Pada tahap ini dilakukan pengecekan struktur dataset untuk mengetahui jumlah data, tipe data, serta memastikan apakah terdapat missing values.

In [6]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857 entries, 0 to 856
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         857 non-null    float64
 1   volatile acidity      857 non-null    float64
 2   citric acid           857 non-null    float64
 3   residual sugar        857 non-null    float64
 4   chlorides             857 non-null    float64
 5   free sulfur dioxide   857 non-null    float64
 6   total sulfur dioxide  857 non-null    float64
 7   density               857 non-null    float64
 8   pH                    857 non-null    float64
 9   sulphates             857 non-null    float64
 10  alcohol               857 non-null    float64
 11  quality               857 non-null    int64  
 12  Id                    857 non-null    int64  
dtypes: float64(11), int64(2)
memory usage: 87.2 KB


### Interpretasi Struktur Data

Berdasarkan output `train.info()`, dapat diketahui bahwa dataset memiliki:

- Jumlah data: 857 baris (entries)
- Jumlah variabel: 13 kolom

Seluruh kolom memiliki jumlah data yang sama yaitu 857 (non-null), sehingga dapat disimpulkan bahwa **tidak terdapat missing values** pada dataset.

Tipe data pada dataset terdiri dari:
- float64 → untuk variabel numerik kontinu (seperti acidity, alcohol, dll)
- int64 → untuk variabel diskrit (quality dan Id)

Variabel `quality` merupakan variabel target yang akan diprediksi, sedangkan `Id` hanya sebagai identitas data dan tidak digunakan dalam proses pemodelan.

Dengan demikian, dataset sudah bersih dan siap untuk tahap analisis selanjutnya.

## 4. Statistik Deskriptif

Tahap ini bertujuan untuk melihat ringkasan statistik dari setiap variabel seperti nilai rata-rata, minimum, maksimum, dan standar deviasi.

In [7]:
train.describe()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
count,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000
mean,8.261960,0.529393,0.267351,2.506184,0.086830,15.782964,45.978413,0.996692,3.313092,0.656709,10.430338,5.653442,813.749125
std,1.701992,0.179162,0.195144,1.293512,0.048721,10.300402,31.692113,0.001901,0.152079,0.167364,1.066971,0.821777,463.807063
min,4.600000,0.120000,0.000000,0.900000,0.012000,1.000000,6.000000,0.990070,2.740000,0.390000,8.400000,3.000000,0.000000
25%,7.100000,0.395000,0.090000,1.900000,0.070000,7.000000,21.000000,0.995520,3.210000,0.550000,9.500000,5.000000,413.000000
50%,7.900000,0.520000,0.250000,2.200000,0.079000,14.000000,38.000000,0.996680,3.310000,0.620000,10.200000,6.000000,814.000000
75%,9.100000,0.640000,0.420000,2.600000,0.090000,21.000000,63.000000,0.997800,3.400000,0.730000,11.100000,6.000000,1214.000000
max,15.600000,1.580000,1.000000,15.500000,0.611000,68.000000,278.000000,1.003200,4.010000,2.000000,14.000000,8.000000,1597.000000


### Interpretasi Statistik Deskriptif

Berdasarkan hasil statistik deskriptif, diperoleh informasi sebagai berikut:

1. Jumlah Data  
Seluruh variabel memiliki jumlah data sebanyak 857, yang menunjukkan tidak adanya data yang hilang.

2. Nilai Rata-rata (Mean)  
Rata-rata dari setiap variabel menunjukkan pusat data. Sebagai contoh:
- fixed acidity memiliki rata-rata sekitar 8.26
- alcohol memiliki rata-rata sekitar 10.43
- quality memiliki rata-rata sekitar 5.65

Hal ini menunjukkan bahwa sebagian besar kualitas anggur berada pada nilai menengah.

3. Sebaran Data (Standar Deviasi)  
Nilai standar deviasi menunjukkan tingkat penyebaran data dari rata-rata.
- Beberapa variabel seperti total sulfur dioxide memiliki standar deviasi yang cukup besar, menunjukkan variasi data yang tinggi.
- Sedangkan variabel seperti density memiliki standar deviasi kecil, sehingga datanya lebih stabil.

4. Nilai Minimum dan Maksimum  
Perbedaan antara nilai minimum dan maksimum cukup besar pada beberapa variabel, seperti:
- total sulfur dioxide (6 – 278)
- residual sugar (0.9 – 15.5)

Hal ini mengindikasikan kemungkinan adanya variasi data yang tinggi atau outlier.

5. Analisis Variabel Target (quality)  
- Nilai minimum = 3
- Nilai maksimum = 8
- Rata-rata ≈ 5.65

Ini menunjukkan bahwa kualitas anggur dalam dataset cenderung berada pada kategori sedang.

Kesimpulan:
Dataset memiliki distribusi nilai yang bervariasi antar fitur, sehingga diperlukan proses normalisasi agar model dapat bekerja lebih optimal.

## 5. Menghapus Kolom yang Tidak Diperlukan

Kolom `Id` tidak digunakan dalam proses pemodelan karena hanya berfungsi sebagai identitas data, sehingga perlu dihapus.

In [8]:
train = train.drop('Id', axis=1)

## 6. Pemisahan Fitur dan Target

Dataset dipisahkan menjadi:
- X (fitur) → variabel independen
- y (target) → variabel yang akan diprediksi yaitu quality

In [9]:
X = train.drop('quality', axis=1)
y = train['quality']

## 7. Normalisasi Data (Feature Scaling)

Dilakukan normalisasi menggunakan StandardScaler agar setiap fitur memiliki skala yang sama, sehingga model dapat bekerja lebih optimal.

In [10]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_scaled = scaler.fit_transform(X)

### Interpretasi Preprocessing

Pada tahap ini dilakukan beberapa proses persiapan data sebelum pemodelan, yaitu:

1. Kolom `Id` dihapus karena tidak berpengaruh terhadap prediksi.
2. Dataset dipisahkan menjadi fitur (X) dan target (y), dimana target yang digunakan adalah `quality`.
3. Dilakukan normalisasi data menggunakan StandardScaler sehingga setiap fitur memiliki skala yang sama.

Hasil normalisasi menunjukkan bahwa nilai data telah berubah menjadi memiliki rata-rata sekitar 0 dan standar deviasi 1.

## 8. Pembagian Data Training dan Testing

Data dibagi menjadi data training dan data testing untuk mengevaluasi performa model.

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

In [12]:
print(X_train.shape)
print(X_val.shape)

(685, 11)
(172, 11)


### Interpretasi Pembagian Data

Berdasarkan hasil pembagian data, diperoleh:

- Data training (X_train): 685 data dengan 11 fitur
- Data validation/testing (X_val): 172 data dengan 11 fitur

Pembagian dilakukan dengan proporsi 80% data training dan 20% data testing.

Data training digunakan untuk melatih model, sedangkan data testing digunakan untuk mengevaluasi performa model.

Jumlah fitur sebanyak 11 menunjukkan bahwa semua variabel independen telah digunakan setelah menghapus kolom `Id` dan memisahkan variabel target `quality`.

Pembagian ini sudah sesuai dan cukup untuk membangun model klasifikasi yang baik.

## 9. Pembuatan Model Klasifikasi

Pada tahap ini digunakan algoritma Random Forest untuk membangun model klasifikasi dalam memprediksi kualitas anggur.

In [13]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(random_state=42)

model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## 10. Prediksi Data Validasi

Model yang telah dilatih digunakan untuk memprediksi data validation.

In [14]:
y_pred = model.predict(X_val)

## 11. Evaluasi Model

Evaluasi dilakukan untuk mengetahui performa model menggunakan metrik akurasi dan confusion matrix.

In [15]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Akurasi:", accuracy_score(y_val, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_val, y_pred))

Akurasi: 0.5813953488372093
Confusion Matrix:
 [[ 0  1  2  0  0]
 [ 0 42 25  0  0]
 [ 0 23 49  6  0]
 [ 0  2 10  9  0]
 [ 0  0  0  3  0]]


### Interpretasi Hasil Evaluasi Model

Berdasarkan hasil evaluasi model Random Forest, diperoleh:

1. Akurasi Model  
Nilai akurasi sebesar 0.5814 atau sekitar 58.14%.  
Hal ini menunjukkan bahwa model mampu memprediksi dengan benar sekitar 58% dari total data.

Akurasi ini tergolong sedang, yang berarti model sudah mampu menangkap pola dalam data, namun masih terdapat kesalahan prediksi yang cukup signifikan.

2. Confusion Matrix  

Confusion matrix menunjukkan distribusi hasil prediksi model terhadap nilai sebenarnya.

- Pada kelas kualitas 5 dan 6, model cukup baik dalam melakukan prediksi, terlihat dari jumlah prediksi benar yang relatif tinggi.
- Namun, pada kelas kualitas yang lebih rendah atau lebih tinggi (seperti 3, 4, dan 7), model kurang mampu memprediksi dengan baik.
- Terlihat banyak kesalahan prediksi antar kelas yang berdekatan, misalnya antara kualitas 5 dan 6, yang menunjukkan bahwa karakteristik data antar kelas tersebut cukup mirip.

3. Analisis Kesalahan Model  
Model cenderung:
- Sulit membedakan kelas dengan jumlah data sedikit (kelas minoritas)
- Lebih sering memprediksi ke kelas tengah (5 dan 6)

Hal ini kemungkinan disebabkan oleh distribusi data yang tidak seimbang (imbalanced dataset).

Kesimpulan:
Model Random Forest yang digunakan sudah mampu melakukan prediksi, namun performanya masih dapat ditingkatkan, terutama dalam membedakan kelas dengan nilai kualitas yang jarang muncul.

##12. Prediksi Data testing

In [19]:
test_processed = test.drop('Id', axis=1)
# scaling data testing
test_scaled = scaler.transform(test_processed)

# prediksi
pred_test = model.predict(test_scaled)

In [20]:
submission = pd.DataFrame({
    'Id': test['Id'],   # pastikan ada kolom Id di test
    'quality': pred_test
})

submission.to_csv('hasilprediksi_123.csv', index=False)

In [21]:
submission.head()

,Id,quality
0,222,5
1,1514,5
2,417,5
3,754,5
4,516,5


In [22]:
submission = pd.DataFrame({
    'Id': test['Id'],
    'quality': pred_test
})

In [23]:
submission.head()

,Id,quality
0,222,5
1,1514,5
2,417,5
3,754,5
4,516,5


In [24]:
submission.to_csv('hasilprediksi_045.csv', index=False)